File: nodejs/generateQRcode.js
const QRCode = require('qrcode')
const { createCanvas, loadImage } = require('canvas')
const fs = require('fs')

async function generateQRCode(text, type, options) {
  const canvasSize = options.canvasSize || 400
  const padding = options.padding || 10
  const bgColor = options.bgColor || '#ffffff'
  const fgColor = options.fgColor || '#000000'
  const borderColor = options.borderColor || '#000000'
  const innerBorderColor = options.innerBorderColor || '#000000'

  let data
  switch (type) {
    case 'text':
      data = text
      break
    case 'tel':
      data = `tel:${text}`
      break
    case 'sms':
      data = `sms:${text}`
      break
    case 'vcard':
      const [name, phone, company, url, title, email, address] = text.split('|')
      data = `BEGIN:VCARD\nVERSION:3.0\nFN:${name}\nORG:${company}\nTITLE:${title}\nTEL;TYPE=WORK,VOICE:${phone}\nADR;TYPE=WORK:${address}\nEMAIL:${email}\nURL:${url}\nEND:VCARD`
      break
    case 'url':
      data = text
      break
    case 'wifi':
      const [ssid, password, security] = text.split('|')
      data = `WIFI:T:${security};S:${ssid};P:${password};;`
      break
    default:
      throw new Error('Unsupported type')
  }

  const canvas = createCanvas(canvasSize, canvasSize)
  const context = canvas.getContext('2d')

  context.fillStyle = bgColor
  context.fillRect(0, 0, canvasSize, canvasSize)

  const qrCodeOptions = {
    margin: 1,
    color: {
      dark: fgColor,
      light: bgColor
    },
    errorCorrectionLevel: 'H'
  }

  const qrCodeData = await QRCode.toDataURL(data, qrCodeOptions)

  const qrCodeImage = await loadImage(qrCodeData)
  context.drawImage(qrCodeImage, padding, padding, canvasSize - padding * 2, canvasSize - padding * 2)

  context.strokeStyle = borderColor
  context.lineWidth = 4
  context.strokeRect(0, 0, canvasSize, canvasSize)

  context.strokeStyle = innerBorderColor
  context.lineWidth = 2
  context.strokeRect(padding, padding, canvasSize - padding * 2, canvasSize - padding * 2)

  const fileName = `${type}_${Date.now()}.png`
  const stream = canvas.createPNGStream()
  const writeStream = fs.createWriteStream(fileName)
  stream.pipe(writeStream)
  return new Promise(resolve => {
    writeStream.on('finish', () => resolve(fileName))
  })
}

(async () => {
  const options = {
    canvasSize: 500,
    padding: 20,
    bgColor: '#ffffff',
    fgColor: '#000000',
    borderColor: '#000000',
    innerBorderColor: '#000000'
  }
  const type = 'text'
  const text = 'Hello World!'
  try {
    const fileName = await generateQRCode(text, type, options)
    console.log("Generated QR code: ${fileName}")
  } catch (error) {
    console.error(error)
  } 
})()

File: python/generateQRcode.py
#生成二维码，参考http://tools.jb51.net/transcoding/jb51qrcode
import qrcode
from PIL import Image

def generate_qrcode(content_type, content, bg_color, img_color, outer_frame_color, inner_frame_color):
    qr = qrcode.QRCode(version=1, box_size=10, border=4)
    if content_type == 'text':
        qr.add_data(content)
    elif content_type == 'tel':
        qr.add_data('tel:' + content)
    elif content_type == 'sms':
        tel, message = content.split(',')
        qr.add_data('smsto:' + tel + ':' + message)
    elif content_type == 'vcard':
        name, tel, company, url, job_title, email, address = content.split(',')
        vcard = f'BEGIN:VCARD\nVERSION:3.0\nN:{name};;;\nFN:{name}\nORG:{company}\nURL:{url}\nTITLE:{job_title}\nTEL;TYPE=WORK,VOICE:{tel}\nEMAIL:{email}\nADR;TYPE=WORK:;;{address};;;;\nEND:VCARD'
        qr.add_data(vcard)
    elif content_type == 'url':
        qr.add_data(content)
    elif content_type == 'wifi':
        ssid, password, encryption = content.split(',')
        wifi = f'WIFI:S:{ssid};T:{encryption};P:{password};;'
        qr.add_data(wifi)
    else:
        raise ValueError('Unsupported content type')
    qr.make(fit=True)
    img = qr.make_image(fill_color=img_color, back_color=bg_color).convert('RGBA')
    outer_frame = Image.new('RGBA', (img.size[0] + 20, img.size[1] + 20), outer_frame_color)
    inner_frame = Image.new('RGBA', (img.size[0] + 16, img.size[1] + 16), inner_frame_color)
    outer_frame.paste(inner_frame, (2, 2))
    outer_frame.paste(img, (10, 10), img)
    return outer_frame


bg_color = '#FFFFFF'
img_color = '#000000'
outer_frame_color = '#000000'
inner_frame_color = '#FFFFFF'

# 生成文本内容的二维码
content_type = 'text'
content = 'Hello, world!'
img = generate_qrcode(content_type, content, bg_color, img_color, outer_frame_color, inner_frame_color)
img.show()

# 生成电话号码的二维码
content_type = 'tel'
content = '1234567890'
img = generate_qrcode(content_type, content, bg_color, img_color, outer_frame_color, inner_frame_color)
img.show()

# 生成短信的二维码
content_type = 'sms'
content = '1234567890,Hello world!'
img = generate_qrcode(content_type, content, bg_color, img_color, outer_frame_color, inner_frame_color)
img.show()

#生成电子名片的二维码
content_type = 'vcard'
content = 'John Doe,1234567890,Acme Inc.,https://www.acme.com,Software Engineer,john.doe@acme.com,123 Main St Anytown USA'
img = generate_qrcode(content_type, content, bg_color, img_color, outer_frame_color, inner_frame_color)
img.show()

#生成网址的二维码
content_type = 'url'
content = 'https://www.example.com'
img = generate_qrcode(content_type, content, bg_color, img_color, outer_frame_color, inner_frame_color)
img.show()

#生成wifi网络的二维码
content_type = 'wifi'
content = 'MyNetwork,MyPassword,WPA'
img = generate_qrcode(content_type, content, bg_color, img_color, outer_frame_color, inner_frame_color)
img.show()